src/compiler/scala/tools/nsc/backend/jvm/analysis/BackendUtils.scala

```diff
@@ -87,15 +100,40 @@ class BackendUtils[BT <: BTypes](val btypes: BT) {
 
     val nilLookupDesc = MethodBType(Nil, jliMethodHandlesLookupRef).descriptor
     val serlamObjDesc = MethodBType(jliSerializedLambdaRef :: Nil, ObjectRef).descriptor
+    val implMethodsArray = implMethods.toArray
 
-    {
-      val mv = cw.visitMethod(ACC_PRIVATE + ACC_STATIC + ACC_SYNTHETIC, "$deserializeLambda$", serlamObjDesc, null, null)
-      mv.visitCode()
+    val mv = cw.visitMethod(ACC_PRIVATE + ACC_STATIC + ACC_SYNTHETIC, "$deserializeLambda$", serlamObjDesc, null, null)
+    def emitLambdaDeserializeIndy(targetMethods: Seq[Handle]) {
       mv.visitVarInsn(ALOAD, 0)
-      mv.visitInvokeDynamicInsn("lambdaDeserialize", serlamObjDesc, lambdaDeserializeBootstrapHandle, implMethods.toArray: _*)
+      mv.visitInvokeDynamicInsn("lambdaDeserialize", serlamObjDesc, lambdaDeserializeBootstrapHandle, targetMethods: _*)
+    }
+
+    val targetMethodGroupLimit = 255 - 1 - 3 // JVM limit. See See MAX_MH_ARITY in CallSite.java
+    val groups: Array[Array[Handle]] = implMethodsArray.grouped(targetMethodGroupLimit).toArray
+    val numGroups = groups.length
+
+    import scala.tools.asm.Label
+    val labelss = Array.fill(numGroups - 1, 3)(new Label())
+    for ((labels, i) <- labelss.iterator.zipWithIndex) {
+      mv.visitTryCatchBlock(labels(0), labels(1), labels(2), "java/lang/IllegalArgumentException")
+    }
+    val terminalLabel = new Label
+    def nextLabel(i: Int) = if (i == numGroups - 2) terminalLabel else labelss(i + 1)(0)
+
+    for ((labels, i) <- labelss.iterator.zipWithIndex) {
+      mv.visitLabel(labels(0))
+      emitLambdaDeserializeIndy(groups(i))
+      mv.visitLabel(labels(1))
       mv.visitInsn(ARETURN)
-      mv.visitEnd()
+      mv.visitLabel(labels(2))
+      mv.visitVarInsn(ASTORE, 1)
+      mv.visitVarInsn(ALOAD, 1)
+      mv.visitMethodInsn(INVOKEVIRTUAL, "java/lang/IllegalArgumentException", "getCause", "()Ljava/lang/Throwable;", false)
+      mv.visitJumpInsn(IFNULL, nextLabel(i))
```

Comment:
Actually, I've just noticed that the second `throw IllegalArgumentException` in `LambdaDeserializer` is detritus that should have been removed in https://github.com/scala/scala/commit/131402fd5fe8c064ef5cfffbe568507cbdf37990.

I'll remove that `throw` and then simplify this PR (no need to check `getCause == null` here, just swallow any `IAE` unconditionally in all but the terminal `indy` call.